src/main/java/coupon/config/db/DataSourceConfig.java

```diff
@@ -0,0 +1,47 @@
+package coupon.config.db;
+
+import java.util.HashMap;
+import java.util.Map;
+import javax.sql.DataSource;
+import jakarta.persistence.EntityManagerFactory;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.DependsOn;
+import org.springframework.context.annotation.Primary;
+import org.springframework.jdbc.datasource.LazyConnectionDataSourceProxy;
+import org.springframework.orm.jpa.JpaTransactionManager;
+import org.springframework.transaction.PlatformTransactionManager;
+
+@Configuration
+public class DataSourceConfig {
+
+    @DependsOn({DataSourceConstants.WRITER_DATA_SOURCE, DataSourceConstants.READER_DATA_SOURCE})
+    @Bean
+    public DataSource routingDataSource(
+            @Qualifier(DataSourceConstants.WRITER_DATA_SOURCE) DataSource writer,
+            @Qualifier(DataSourceConstants.READER_DATA_SOURCE) DataSource reader
+    ) {
+        DynamicRoutingDataSource routingDataSource = new DynamicRoutingDataSource();
+        Map<Object, Object> dataSourceMap = new HashMap<>();
+        dataSourceMap.put(DataSourceConstants.WRITER, writer);
+        dataSourceMap.put(DataSourceConstants.READER, reader);
+        routingDataSource.setTargetDataSources(dataSourceMap);
+        routingDataSource.setDefaultTargetDataSource(writer);
+        return routingDataSource;
+    }
+
+    @Primary
+    @DependsOn({"routingDataSource"})
+    @Bean
+    public DataSource dataSource(DataSource routingDataSource) {
+        return new LazyConnectionDataSourceProxy(routingDataSource);
```

Comment:
LazyConnectionDataSourceProxy 는 무엇이고 어떤 역할을 할까요?

Reply:
>결론적으로 `@Primary` 애노테이션을 활용해서 `DataSource` 타입의 의존성을 주입받을 때 `LazyConnectionDataSourceProxy`가 반환되도록 설정하여 `Connection`을 받아오는 시점을 지연시킵니다.

결론만을 말씀드리는 것보다 아래 제가 정리한 내용을 말씀드리는게 더 낫다고 생각하여 아래에 첨부했습니다!
틀린 부분이나 잘못된 부분이 있으면 말씀 부탁드릴게요 🙇‍♂️

![tran1](https://github.com/user-attachments/assets/a0f7157e-6d4c-4818-a0ea-a59f77dffbde)

- Spring은 트랜잭션 시작 시점에 TransactionManger를 통해 Connection을 받아와서 TransactionSynchronizationManger에 Connection을 보관
- Connection은 Thead Local한 TransactionSynchronizationManager의 ConnectionHolder 에 저장. 이후 데이터 접근 로직이 수행될 때 Connection을 받아 실제 데이터베이스 서버에 요청을 전송. 이를 통해 하나의 요청이 같은 Connection을 사용할 수 있도록 보장

> 스프링 프레임워크에서 `TransactionManager`는 여러 트랜잭션 관리 인터페이스를 지원하며, 각 기술(JDBC, JPA, Hibernate 등)에 맞는 트랜잭션 관리 방식을 제공합니다. 스프링에서는 주로 **`PlatformTransactionManager`** 인터페이스를 구현한 클래스들이 트랜잭션을 관리
> 
> 
> ### `TransactionManager`의 역할
> 
> - **트랜잭션 시작**: 트랜잭션을 시작하고, 해당 트랜잭션 내에서 수행될 작업을 관리합니다.
> - **커밋**: 트랜잭션이 성공적으로 완료되었을 때 모든 작업을 커밋하여 변경 사항을 영구적으로 반영합니다.
> - **롤백**: 트랜잭션 중 오류가 발생하거나 중간에 실패하면 롤백을 수행하여 이전 상태로 되돌립니다.

<img width="547" alt="tran2" src="https://github.com/user-attachments/assets/de0dac96-78b7-4738-9516-55559477ff27">

<img width="989" alt="tran3" src="https://github.com/user-attachments/assets/af7430a2-fc78-4739-a51f-64a0a3155711">

**`PlatformTransactionManager`** 인터페이스를 구현한 클래스 예시로는 아래와 같은 것들이 있습니다

- DataSourceTransactionManager
- JpaTransactionManager
- HibernateTransactionManager
- JtaTransactionManager

<img width="680" alt="tran4" src="https://github.com/user-attachments/assets/3d84ab8f-2032-471a-ad05-4bb6836bed90">


JPA 기반 애플리케이션에서는 트랜잭션 관리가 중요한 역할을 하기 때문에, 트랜잭션 관리자가 필요합니다. 여기서는 `JpaTransactionManager`가 이를 담당합니다.

이 경우에 `@Transactional`이 걸려있는 메서드를 호출하게 된다면 `JpaTransactionManager`가 트랜잭션을 처리하기 위한 사전작업을 아래와 같은 순서로 수행됩니다.

> JpaTransactionManager
> 

<img width="1255" alt="tran5" src="https://github.com/user-attachments/assets/a9ef67bc-cb35-4149-8379-dd2ac22f6d33">

`AbstractPlatformTransactionManager`의 `startTransaction`에서 구현체의 `doBegin()`, `prepareSynchronization()` 메서드를 호출

> `prepareSynchronization()`
> 

<img width="890" alt="tran6" src="https://github.com/user-attachments/assets/41a1f697-7f0c-4380-a8d9-bc897814c456">

이 메서드는 새로운 트랜잭션이 시작될 때 트랜잭션과 관련된 정보(트랜잭션의 이름, 격리 수준, 읽기 전용 여부 등)를 `TransactionSynchronizationManager`에 설정하여, 트랜잭션 내에서 올바르게 동작할 수 있도록 준비합니다.

```java
// JpaTransactionManager
    protected void doBegin(Object transaction, TransactionDefinition definition) {
        JpaTransactionObject txObject = (JpaTransactionObject)transaction;
        if (txObject.hasConnectionHolder() && !txObject.getConnectionHolder().isSynchronizedWithTransaction()) {
            throw new IllegalTransactionStateException("Pre-bound JDBC Connection found! JpaTransactionManager does not support running within DataSourceTransactionManager if told to manage the DataSource itself. It is recommended to use a single JpaTransactionManager for all transactions on a single DataSource, no matter whether JPA or JDBC access.");
        } else {
            try {
                EntityManager em;
                if (!txObject.hasEntityManagerHolder() || txObject.getEntityManagerHolder().isSynchronizedWithTransaction()) {
                    em = this.createEntityManagerForTransaction();
                    if (this.logger.isDebugEnabled()) {
                        this.logger.debug("Opened new EntityManager [" + em + "] for JPA transaction");
                    }

                    txObject.setEntityManagerHolder(new EntityManagerHolder(em), true);
                }

                em = txObject.getEntityManagerHolder().getEntityManager();
                int timeoutToUse = this.determineTimeout(definition);
                Object transactionData = this.getJpaDialect().beginTransaction(em, new JpaTransactionDefinition(definition, timeoutToUse, txObject.isNewEntityManagerHolder()));
                txObject.setTransactionData(transactionData);
                txObject.setReadOnly(definition.isReadOnly());
                if (timeoutToUse != -1) {
                    txObject.getEntityManagerHolder().setTimeoutInSeconds(timeoutToUse);
                }

                if (this.getDataSource() != null) {
                    ConnectionHandle conHandle = this.getJpaDialect().getJdbcConnection(em, definition.isReadOnly());
                    if (conHandle != null) {
                        ConnectionHolder conHolder = new ConnectionHolder(conHandle);
                        if (timeoutToUse != -1) {
                            conHolder.setTimeoutInSeconds(timeoutToUse);
                        }

                        if (this.logger.isDebugEnabled()) {
                            this.logger.debug("Exposing JPA transaction as JDBC [" + conHandle + "]");
                        }

                        TransactionSynchronizationManager.bindResource(this.getDataSource(), conHolder);
                        txObject.setConnectionHolder(conHolder);
                    } else if (this.logger.isDebugEnabled()) {
                        this.logger.debug("Not exposing JPA transaction [" + em + "] as JDBC transaction because JpaDialect [" + this.getJpaDialect() + "] does not support JDBC Connection retrieval");
                    }
                }

                if (txObject.isNewEntityManagerHolder()) {
                    TransactionSynchronizationManager.bindResource(this.obtainEntityManagerFactory(), txObject.getEntityManagerHolder());
                }

                txObject.getEntityManagerHolder().setSynchronizedWithTransaction(true);
            } catch (TransactionException var9) {
                TransactionException ex = var9;
                this.closeEntityManagerAfterFailedBegin(txObject);
                throw ex;
            } catch (Throwable var10) {
                Throwable ex = var10;
                this.closeEntityManagerAfterFailedBegin(txObject);
                throw new CannotCreateTransactionException("Could not open JPA EntityManager for transaction", ex);
            }
        }
    }
```
---

위의 doBegin()메서드는 많은 역할을 하지만 간략히 설명하면, EntityManager를 생성하고, 
JPA 트랜잭션이 JDBC 연결을 필요로 하는 경우, `ConnectionHandle`을 통해 JDBC 연결을 얻고 이를 트랜잭션 컨텍스트에 바인딩합니다. 새로 생성된 `EntityManager`와 `EntityManagerFactory`를 `TransactionSynchronizationManager`에 바인딩하여 트랜잭션과 동기화되도록 설정합니다.

여기서 집중해야 할 부분은 아래의 코드에 해당합니다

```java
if (this.getDataSource() != null) {
    // JPA 트랜잭션에서 JDBC Connection을 가져오고, 트랜잭션이 읽기 전용인지 여부를 설정
    ConnectionHandle conHandle = this.getJpaDialect().getJdbcConnection(em, definition.isReadOnly());
    
    if (conHandle != null) {
        // 가져온 Connection을 ConnectionHolder에 감싸서 관리
        ConnectionHolder conHolder = new ConnectionHolder(conHandle);
        
        // 타임아웃이 설정되어 있는 경우, 타임아웃 값을 ConnectionHolder에 설정
        if (timeoutToUse != -1) {
            conHolder.setTimeoutInSeconds(timeoutToUse);
        }

        // 디버그 로그로 JDBC Connection을 JPA 트랜잭션에 노출한다는 내용 출력
        if (this.logger.isDebugEnabled()) {
            this.logger.debug("Exposing JPA transaction as JDBC [" + conHandle + "]");
        }

        // TransactionSynchronizationManager에 DataSource와 ConnectionHolder를 바인딩하여 동기화
        TransactionSynchronizationManager.bindResource(this.getDataSource(), conHolder);
        
        // 트랜잭션 객체에 ConnectionHolder 설정
        txObject.setConnectionHolder(conHolder);
    } else if (this.logger.isDebugEnabled()) {
        // 만약 JpaDialect가 JDBC Connection을 지원하지 않는 경우 로그 출력
        this.logger.debug("Not exposing JPA transaction [" + em + "] as JDBC transaction because JpaDialect [" + this.getJpaDialect() + "] does not support JDBC Connection retrieval");
    }
}
```

`getJdbcConnection`을 통해 JPA 트랜잭션 내에서 JDBC `Connection`을 얻고, 이때 트랜잭션의 `readOnly` 여부를 반영하기에 트랜잭션의 readOnly속성에 따라 Connection을 분기해서 사용할려면 Connection을 가져오는 시점을 뒤로 미루어야 합니다.

LazyConnectionDataSourceProxy가 반환되도록 설정하지 않은 경우에는 아래 순서로 실행되고,

- @Transactional이 시작되면 JpaTransactionManager가 영속성 컨텍스트를 생성하고
- Connection을 가져와 **`TransactionSynchronizationManager.bindResource()`** 이 메서드를 통해 현재 스레드(`ThreadLocal`)에 Connection을 바인딩
- `EntityManagerHolder` 객체가 트랜잭션과 동기화되었음을 표시하는 `txObject.getEntityManagerHolder().setSynchronizedWithTransaction(true);` 메서드를 호출합니다. → `EntityManager`가 현재 트랜잭션과 제대로 동기화되었음을 확인
- 서비스 내부 로직을 수행하면서 저장된 Connection을 사용

최종적으로 TransactionManager에 의해 transaction이 commit

LazyConnectionDataSourceProxy가 반환되도록 설정한 경우에는 아래 순서로 실행됩니다.

- `@Transactional`이 시작되면 JpaTransactionManager가 영속성 컨텍스트를 생성하고
- Connection을 가져와 **`TransactionSynchronizationManager.bindResource()`** 이 메서드를 통해 현재 스레드(`ThreadLocal`)에 Connection을 바인딩
→ 이때 `ProxyConnection`을 받고 트랜잭션 동기화를 진행
    
<img width="1002" alt="tran7" src="https://github.com/user-attachments/assets/160ed47f-f840-4714-bb46-25fb6ad20845">
    
- `EntityManagerHolder` 객체가 트랜잭션과 동기화되었음을 표시하는 `txObject.getEntityManagerHolder().setSynchronizedWithTransaction(true);` 메서드를 호출합니다. → `EntityManager`가 현재 트랜잭션과 제대로 동기화되었음을 확인
- 서비스 내부 로직을 수행하면서 저장된 Connection을 사용

이후 쿼리가 수행되는 시점에 (실제 Connection이 필요한 시점) 적합한 Datasource로부터 Connection을 받아오게 됩니다.

<img width="1042" alt="tran8" src="https://github.com/user-attachments/assets/91416445-d171-4b9c-97db-75d5d6e6bc98">

위의 `targetDataSource`로 `routingDataSource` 가 사용됩니다.

`isCurrentTransactionReadOnly` 메서드를 사용하여 현재 트랜잭션이 `readOnly`속성인지 아닌지를 분기합니다.

```java
    @DependsOn({WRITER_DATA_SOURCE, READER_DATA_SOURCE})
    @Bean
    public DataSource routingDataSource(
            @Qualifier(WRITER_DATA_SOURCE) DataSource writer,
            @Qualifier(READER_DATA_SOURCE) DataSource reader
    ) {
        DynamicRoutingDataSource routingDataSource = new DynamicRoutingDataSource();
        Map<Object, Object> dataSourceMap = new HashMap<>();
        dataSourceMap.put(WRITER, writer);
        dataSourceMap.put(READER, reader);
        routingDataSource.setTargetDataSources(dataSourceMap);
        routingDataSource.setDefaultTargetDataSource(writer);
        return routingDataSource;
    }

public class DynamicRoutingDataSource extends AbstractRoutingDataSource {
    @Override
    protected Object determineCurrentLookupKey() {
        if (isCurrentTransactionReadOnly()) {
            return READER;
        }
        return WRITER;
    }
}
```

**최종적으로 TransactionManager에 의해 transaction이 commit하게 됩니다**

Reply:
> 이후 쿼리가 수행되는 시점에 (실제 Connection이 필요한 시점) 적합한 Datasource로부터 Connection을 받아오게 됩니다.

열심히 공부해주셨네요 최고입니다 👍👍👍